src/Common/HashTable/PackedHashMap.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <Common/HashTable/HashMap.h>
+
+/// A pair that does not initialize the elements, if not needed.
+///
+/// NOTE: makePairNoInit() is omitted for PackedPairNoInit since it is not
+/// required for PackedHashMap (see mergeBlockWithPipe() for details)
+template <typename First, typename Second>
+struct PackedPairNoInit
```

Comment:
I was building locally and got this error:
```
In file included from /home/ubuntu/repo/ch4/src/Dictionaries/HashedDictionaryCollectionType.h:4:                                                                                                                               
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<128, unsigned int>>' [-Werror,-Wpacked]                             
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<256, unsigned int>>' [-Werror,-Wpacked]                             
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<128, int>>' [-Werror,-Wpacked]                                      
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, wide::integer<256, int>>' [-Werror,-Wpacked]                                      
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::Decimal<long>>' [-Werror,-Wpacked]                                            
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::Decimal<wide::integer<128, int>>>' [-Werror,-Wpacked]                         
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::Decimal<wide::integer<256, int>>>' [-Werror,-Wpacked]                         
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, DB::DateTime64>' [-Werror,-Wpacked]                                               
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:13:12: error: not packing field 'second' as it is non-POD for the purposes of layout [-Werror,-Wpacked-non-pod]                                                     
    Second second;                                                                                                                                                                                                             
           ^                                                                                                                                                                                                                   
/home/ubuntu/repo/ch4/src/Common/HashTable/PackedHashMap.h:10:8: error: packed attribute is unnecessary for 'PackedPairNoInit<unsigned long, StrongTypedef<wide::integer<128, unsigned int>, DB::UUIDTag>>' [-Werror,-Wpacked] 
struct PackedPairNoInit                                                                                                                                                                                                        
       ^                                                                                                                                                                                                                       
fatal error: too many errors emitted, stopping now [-ferror-limit=]                                                                                                                                                            
20 errors generated.                                                                                                                                                                                                           
[46/1294] Building CXX object src/Parsers/CMa...rsers.dir/ASTCreateNamedCollectionQuery.cpp.o                                                                                                                                  
ninja: build stopped: subcommand failed.                                                                                                                                                                                       
```

Reply:
>I was building locally and got this error:

Yeah, I converted to clang-16 too early :)
The problem is that the interpretation of POD types (in terms of layout) had been changed since clang-16 - https://github.com/llvm/llvm-project/commit/277123376ce08c98b07c154bf83e4092a5d4d3c6

Will disable using of PackedHashMap for some types.

Reply:
[`06e8bd6` (#49380)](https://github.com/ClickHouse/ClickHouse/pull/49380/commits/06e8bd63b013d19b67548e99f3e5731226439e9f)